File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.py
def solution(cards):
    answer = []
    visited = [False] * len(cards)
    
    for i in range(len(cards)):
        if not visited[i]:
            visited[i] = True
            cnt = 1
            t = i
            
            while not visited[cards[t] - 1]:
                cnt += 1
                t = cards[t] - 1
                visited[t] = True
            
            answer.append(cnt)
    
    if len(answer) == 1:
        return 0
    
    answer.sort(reverse=True)
    return answer[0] * answer[1]

File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.java
import java.util.*;

class Solution {
    
    public int solution(int[] cards) {
        List<Integer> groups = new ArrayList<>();
        boolean[] visited = new boolean[cards.length];

        for (int i = 0; i < cards.length; i++) {
            if (visited[i]) continue;

            visited[i] = true;
            int count = 1;
            int next = i;

            while (!visited[cards[next] - 1]) {
                next = cards[next] - 1;
                visited[next] = true;
                count++;
            }

            groups.add(count);
        }

        if (groups.size() < 2) return 0;

        Collections.sort(groups, Collections.reverseOrder());
        return groups.get(0) * groups.get(1);
    }
}